rust-symcrypt/src/hmac.rs

```diff
@@ -1,180 +1,318 @@
 use core::ffi::c_void;
 use std::mem;
 use std::ptr;
+use std::pin::Pin;
 use symcrypt_sys;
 
+use crate::errors::SymCryptError;
+
 pub const SHA256_HMAC_RESULT_SIZE: usize = symcrypt_sys::SYMCRYPT_SHA256_RESULT_SIZE as usize;
 pub const SHA384_HMAC_RESULT_SIZE: usize = symcrypt_sys::SYMCRYPT_SHA384_RESULT_SIZE as usize;
 
+/// Generic trait for HMACs
+pub trait Hmac {
+    type Result;
+
+    fn append(&mut self, data: &[u8]);
+    fn result(self) -> Self::Result;
+    fn copy(&self) -> Box<Self>;
+}
+
+pub trait ExpandedKey {
+    type Result;
+
+    fn new() -> Self;
+    fn expand(&mut self, key: &[u8]) -> Self::Result;
+}
+
+pub struct HmacSha256ExpandedKey(symcrypt_sys::SYMCRYPT_HMAC_SHA256_EXPANDED_KEY);
+
+impl ExpandedKey for HmacSha256ExpandedKey {
+    type Result = symcrypt_sys::SYMCRYPT_HMAC_SHA256_EXPANDED_KEY;
+
+    fn new() -> Self {
+        HmacSha256ExpandedKey(symcrypt_sys::SYMCRYPT_HMAC_SHA256_EXPANDED_KEY::default())
+    }
+
+    fn expand(&mut self, key: &[u8]) -> Self::Result {
+        unsafe {
+            symcrypt_sys::SymCryptHmacSha256ExpandKey(
+                &mut self.0,
+                key.as_ptr(),
+                key.len() as symcrypt_sys::SIZE_T,
+            );
+        }
+        self.0
+    }
+}
+
 pub struct HmacSha256State {
+    inner: Pin<Box<HmacSha256Inner>>,
+}
+
+struct HmacSha256Inner {
     state: symcrypt_sys::SYMCRYPT_HMAC_SHA256_STATE,
     expanded_key: symcrypt_sys::SYMCRYPT_HMAC_SHA256_EXPANDED_KEY,
```

Comment:
>symcrypt_sys::SYMCRYPT_HMAC_SHA256_EXPANDED_KEY, [](http://example.com/codeflow?start=18&length=48)

This will need to Rc<> or Arc<> 
 For when used in the copy().

Reply:
Also changed to heap allocaation pointer.